src/kernel/txgraph.cpp

```diff
@@ -0,0 +1,1353 @@
+            tx_size += tx.get().GetTxSize();
+        }
+    }
+    if (tx_count != m_tx_count) return false;
+    if (tx_size != m_tx_size) return false;
+
+    // Check topology.
+    std::set<TxEntry::TxEntryRef, TxEntry::CompareById> seen_elements;
+    for (auto &chunk : m_chunks) {
+        for (auto tx : chunk.txs) {
+            for (auto parent : tx.get().GetTxEntryParents()) {
+                if (seen_elements.count(parent) == 0) return false;
+            }
+            seen_elements.insert(tx);
+            if (tx.get().m_cluster != this) return false;
+            if (&m_chunks[tx.get().m_loc.first] != &chunk) return false;
+            if (tx.get().m_loc.second->get().unique_id != tx.get().unique_id) return false;
+        }
+    }
+    return true;
+}
+
+bool TxGraphCluster::CheckTopo() const
+{
+    // Check topology.
+    std::set<TxEntry::TxEntryRef, TxEntry::CompareById> seen_elements;
+    for (auto &chunk : m_chunks) {
+        for (auto tx : chunk.txs) {
+            for (auto parent : tx.get().GetTxEntryParents()) {
+                if (seen_elements.count(parent) == 0) return false;
+            }
+            seen_elements.insert(tx);
+        }
+    }
+    return true;
+}
+
+namespace {
+
+template <typename SetType>
+std::vector<TxEntry::TxEntryRef> InvokeSort(size_t tx_count, const std::vector<TxGraphCluster::Chunk>& chunks)
+{
+    std::vector<TxEntry::TxEntryRef> txs;
+    cluster_linearize::Cluster<SetType> cluster;
+    const auto time_1{SteadyClock::now()};
+
+    std::vector<TxEntry::TxEntryRef> orig_txs;
+    std::vector<std::pair<const TxEntry*, unsigned>> entry_to_index;
+
+    cluster.reserve(tx_count);
+    entry_to_index.reserve(tx_count);
+    cluster.clear();
+    for (auto &chunk : chunks) {
+        for (auto tx : chunk.txs) {
+            orig_txs.emplace_back(tx);
+            cluster.emplace_back(FeeFrac(tx.get().GetModifiedFee(), tx.get().GetTxSize()), SetType{});
+            entry_to_index.emplace_back(&(tx.get()), cluster.size() - 1);
+        }
+    }
+    std::sort(entry_to_index.begin(), entry_to_index.end());
+    for (size_t i=0; i<orig_txs.size(); ++i) {
+        for (auto& parent : orig_txs[i].get().GetTxEntryParents()) {
+            auto it = std::lower_bound(entry_to_index.begin(), entry_to_index.end(), &(parent.get()),
+                    [&](const auto& a, const auto& b) { return std::less<const TxEntry*>()(a.first, b); });
+            assert(it != entry_to_index.end());
+            assert(it->first == &(parent.get()));
+            cluster[i].second.Set(it->second);
+        }
+    }
+    uint64_t iterations = MAX_ITERATIONS;
+
+    std::vector<unsigned int> orig_linearization;
+    orig_linearization.reserve(tx_count);
+    for (unsigned int i=0; i<cluster.size(); ++i) {
+        orig_linearization.push_back(i);
+    }
+    cluster_linearize::DepGraph dep_graph(cluster);
+    auto result = cluster_linearize::Linearize(dep_graph, iterations, 0, orig_linearization);
```

Comment:
Fixed this to use `result.second`, which indicates whether the linearization is optimal.

Also added a wip commit to track iterations done, so that the benchmark logging is more useful for now.